doc/Yang Model Guidelines.md

```diff
@@ -0,0 +1,253 @@
+#define IP_TYPE_NON_IPv4 "NON_IPv4"
+#define IP_TYPE_IPv6ANY "IPV6ANY"
+#define IP_TYPE_NON_IPv6 "NON_IPv6"
+#define IP_TYPE_ARP "ARP"
+#define IP_TYPE_ARP_REQUEST "ARP_REQUEST"
+#define IP_TYPE_ARP_REPLY "ARP_REPLY"
+```
+Example of When Statement: Orchagent of Sonic will have unknown behavior if below config is entered, So yang must have a constraint. Here SRC_IP is IPv4, where as IP_TYPE is IPv6.
+
+#### ABNF:
+```
+   "ACL_RULE": {
+        "NO-NSW-PACL-V4|Rule_20": {
+            "PACKET_ACTION": "FORWARD",
+            "DST_IP": "10.186.72.0/26",
+            "SRC_IP": "10.176.0.0/15",
+            "PRIORITY": "999980",
+            "IP_TYPE": "IPv6"
+        },
+ 
+```
+#### Yang:
+```
+choice ip_prefix {
+                case ip4_prefix {
+                    when "boolean(IP_TYPE[.='ANY' or .='IP' or .='IPV4' or .='IPV4ANY' or .='ARP'])";
+                    leaf SRC_IP {
+                        type inet:ipv4-prefix;
+                    }
+                    leaf DST_IP {
+                        type inet:ipv4-prefix;
+                    }
+                }
+                case ip6_prefix {
+                    when "boolean(IP_TYPE[.='ANY' or .='IP' or .='IPV6' or .='IPV6ANY'])";
+                    leaf SRC_IPV6 {
+                        type inet:ipv6-prefix;
+                    }
+                    leaf DST_IPV6 {
+                        type inet:ipv6-prefix;
+                    }
+                }
+            }
+```
+Example of Pattern: If PORT Range should be "<0-65365> - <0-65365>"
+```
+leaf L4_DST_PORT_RANGE {
+    type string {
+    pattern '([0-9]{1,4}|[0-5][0-9]{4}|[6][0-4][0-9]{3}|[6][5][0-2][0-9]{2}|[6][5][3][0-5]{2}|[6][5][3][6][0-5])-([0-9]{1,4}|[0-5][0-9]{4}|[6][0-4][0-9]{3}|[6][5][0-2][0-9]{2}|[6][5][3][0-5]{2}|[6][5][3][6][0-5])';
+        }
+}
+```
+### 8. Comment all must, when and patterns conditions.
+Example:
+```
+leaf family {
+                /* family leaf needed for backward compatibility
+                   Both ip4 and ip6 address are string in IETF RFC 6020, 
+                   so must statement can check based on : or ., family 
+                   should be IPv4 or IPv6 according.
+                */
+                must "(contains(../ip-prefix, ':') and current()='IPv6') or
+                      (contains(../ip-prefix, '.') and current()='IPv4')";
+                type sonic-head:ip-family;
+            }
+```
+
+
+
+### 9.  Use IETF data types for leaf type first if applicable (RFC 6021) . Declare new type (say Sonic types) only if IETF type is not applicable. All Sonic Types must be part of same header type or common yang model.
+Example:
+```
+                    leaf SRC_IP {
+                        type inet:ipv4-prefix; <<<<
+                    }
+
+                    leaf DST_IP {
+                        type inet:ipv4-prefix;
+                    }
+```
+
+
+## All above steps are manual right now, in future few of them can be automated.
```

Comment:
This needs to be clarified. Are we saying that ABNF to YANG schema translation will be automated to some extend ?

Reply:
I don't get it either. This is a one time task. In future, when we add/update/remove, we work on the YANG, which is a 1:1 representation of ABNF schema.

The CVL YANG is such that it is just a representation of ABNF schema, to the extent, it should be feasible to write a simple tool to get ABNF schema from YANG. Other way round may not be possible, as YANG could have constraints laid out along with some metadata to meet mgmt FW requirements.

Reply:
Can we please drop this comments as we all agree it is manual for now ?
Any future enhancement ideas, let us hold back, until we bake the first cut.